### PR TITLE
CPLAT-12205 Add WithTransition wrapper component to enable "controlled" transitions using props

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
## Motivation
Currently, `AbstractTransitionComponent`s can only be programmatically shown/hidden using the `show` / `hide` / `toggle` API methods on the instance.

With React 16 componentry - this is not ideal since calling the methods from a parent component requires the use of a `ref`, and the child component is no longer guaranteed to have been mounted when the parent component is. This makes it very difficult to hide/show things within data-driven applications without having to work around runtime race conditions.

## Changes
1. Add the `WithTransition` wrapper component. 
    This component is designed to have identical behavior to `AbstractTransitionComponent`, but allows the consumer to control the behavior using the value of `props.isShown`.
2. Add tests.

#### Release Notes
Add `WithTransition` wrapper component to enable "controlled" transitions using props

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/component/abstract_transition_props.dart [hunk 1]
```diff
@@ -25,6 +25,13 @@ mixin TransitionPropsMixin on UiProps {
   static final TransitionPropsMixin defaultProps = TransitionPropsMapView({})
     ..transitionCount = 1;
 
+  /// Prop that allows you to control the visibility of the component with props.
+  ///
+  /// __NOTE:__ In order for this work properly, if you have an implementation
+  /// of [AbstractTransitionComponent.initiallyShown] in your subclass - it must
+  /// use/reference this value.
+  bool shown;
```
Comment (aaronlademann-wf) [inline lib/src/component/abstract_transition_props.dart:33]: Named this way instead of `isShown` to avoid gross prop namespacing / forwarding issues in consumer libs that have an `isShown` prop in a mixin that is used in props classes that also mix in `TransitionPropsMixin`.

File: test/over_react/component/abstract_transition_test.dart [hunk 2]
```diff
@@ -31,613 +32,743 @@ part 'abstract_transition_test.over_react.g.dart';
...
-    group('renders with correct initial state when the `initiallyShown` getter returns', () {
-      test('true', () {
-        var renderedInstance = render(Transitioner()
-          ..initiallyShown = true
-        );
-        TransitionerComponent transitioner = getDartComponent(renderedInstance);
-
-        expect(transitioner.state.transitionPhase, equals(TransitionPhase.SHOWN));
+    @isTestGroup
+    void sharedTransitionerTests(BuilderOnlyUiFactory<AbstractSharedTransitionerProps> factory, {
```
Comment (aaronlademann-wf) [inline test/over_react/component/abstract_transition_test.dart:36]: Strongly recommend reviewing this file using a [whitespace-agnostic diff](https://github.com/Workiva/over_react/pull/621/files?w=1)

**Review by joebingham-wk — COMMENTED** [review]
This all looks awesome! I just have one question, but it's a small one, and then I'm +1!

File: lib/src/component/abstract_transition.dart [hunk 3]
```diff
@@ -100,37 +100,95 @@ abstract class AbstractTransitionComponent<T extends AbstractTransitionProps,
...
+    return (newState()
+      ..transitionPhase = this.initiallyShown ? TransitionPhase.SHOWN : TransitionPhase.HIDDEN
+    );
+  }
+
+  @mustCallSuper
+  @override
+  Map getDerivedStateFromProps(Map nextProps, Map prevState) {
+    final tNextProps = typedPropsFactory(nextProps);
+    if (tNextProps.shown == null) return prevState;
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/abstract_transition.dart:113]: @joebingham-wk on initial mount, `getDerivedStateFromProps` is called - and the `prevState` value is the return value of the `initialState` getter.  So if you return `null`, the `initialState` value(s) will not be applied for the first render.

[the remark this answers is not in the document]
Comment (joebingham-wk) [reply lib/src/component/abstract_transition.dart:113]: @aaronlademann-wf Oh, I didn't know that was expected. The React docs (which I've criticized for not being verbose enough) don't mention that though, and when I run [this reduced test case](https://gist.github.com/joebingham-wk/5ca1fa2492c9927d3fdaa5f7b60a908c), I'm seeing the initial state value come through... I'm sure either way this code works, I just want to make sure I understand the lifecycle method correctly. Does that test case miss something though?

File: lib/src/component/abstract_transition.dart [hunk 4]
```diff
@@ -100,37 +100,95 @@ abstract class AbstractTransitionComponent<T extends AbstractTransitionProps,
   );
 
   @override
-  get initialState => (newState()
-    ..transitionPhase = this.initiallyShown ? TransitionPhase.SHOWN : TransitionPhase.HIDDEN
-  );
+  get initialState {
```
Comment (greglittlefield-wf) [inline lib/src/component/abstract_transition.dart:103]: #nit this could stay as an arrow function

File: lib/src/component/abstract_transition.dart [hunk 5]
```diff
@@ -100,37 +100,95 @@ abstract class AbstractTransitionComponent<T extends AbstractTransitionProps,
...
+
+  @mustCallSuper
+  @override
+  Map getDerivedStateFromProps(Map nextProps, Map prevState) {
+    final tNextProps = typedPropsFactory(nextProps);
+    if (tNextProps.shown == null) return prevState;
+
+    final tPrevState = typedStateFactory(prevState);
+    if (tNextProps.shown && _isOrWillBeHidden(tPrevState)) {
+      return _handleShow(tNextProps, tPrevState, this);
```
Comment (greglittlefield-wf) [inline lib/src/component/abstract_transition.dart:117]: `this` cannot be used in `getDerivedStateFromProps`

File: lib/src/component/abstract_transition.dart [hunk 6]
```diff
@@ -100,37 +100,95 @@ abstract class AbstractTransitionComponent<T extends AbstractTransitionProps,
...
+
   /// Whether transitions are enabled for this component.
-  bool get hasTransition => true;
+  ///
+  /// Will not work when the component is controlled using [TransitionPropsMixin.shown].
+  /// Use [TransitionPropsMixin.transitionCount] instead.
+  ///
+  /// __Deprecated.__ Use [TransitionPropsMixin.transitionCount] instead.
+  @Deprecated('Use `props.transitionCount` instead.')
+  bool get hasTransition => _hasTransition(props);
```
Comment (greglittlefield-wf) [inline lib/src/component/abstract_transition.dart:146]: Same question as below; what's the migration path? https://sourcegraph.wk-dev.wdesk.org/search?q=get%5CshasTransition+repo:%5Egithub%5C.com/Workiva/web_skin_dart%24+&patternType=regexp#1

File: lib/src/component/abstract_transition.dart [hunk 7]
```diff
@@ -311,9 +380,21 @@ abstract class AbstractTransitionComponent<T extends AbstractTransitionProps,
...
+  @Deprecated('Use props.onWillShow instead.')
   void prepareShow() {}
 
   /// Method that will be called right before the [AbstractTransitionComponent] begins to hide.
+  ///
+  /// Will not work when the component is controlled using [TransitionPropsMixin.shown].
+  /// Use [TransitionPropsMixin.onWillHide] instead.
+  ///
+  /// __Deprecated.__ Use [TransitionPropsMixin.onWillHide] instead.
+  @Deprecated('Use props.onWillHide instead.')
```
Comment (greglittlefield-wf) [inline lib/src/component/abstract_transition.dart:397]: Do we have a migration path for WSD components consuming these APIs? https://sourcegraph.wk-dev.wdesk.org/search?q=prepare%28Hide%7CShow%29+repo:%5Egithub%5C.com/Workiva/web_skin_dart%24+&patternType=regexp#3

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Given the `getDerivedStateFromProps` issues and deprecation concerns, is it possible to add controlled behavior to this component cleanly?

We might be better off starting fresh with a new `TransitionPrimitive` component that's entirely props-based

File: test/over_react/component/with_transition_test.dart [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+import 'package:over_react/over_react.dart' hide TransitionPropsMixin, $TransitionPropsMixin, AbstractTransitionProps, AbstractTransitionState, AbstractTransitionComponent;
+import 'package:over_react/components.dart';
+import 'package:over_react_test/over_react_test.dart';
+import 'package:test/test.dart';
+
+import '../../test_util/test_util.dart';
+
+part 'with_transition_test.over_react.g.dart';
+
+/// Main entry point for AbstractTransition testing
```
Comment (joebingham-wk) [inline test/over_react/component/with_transition_test.dart:30]: #nit
```suggestion
/// Main entry point for WithTransition testing
```

**Review by joebingham-wk — DISMISSED** [review]
+1! This all looks great to me, and shows a lot of cool things you can do with functional components, hooks, along with how to test for them!

**Review by maxwellpeterson-wf — DISMISSED** [review]
This is working well in two different places in my shell refactor branch. The documentation is also excellent and easy to follow!

**Comment by maxwellpeterson-wf** [comment]
+10
- Able to use `childPropsByPhase` to set various props on the child component
- Event callbacks (such as `onDidHide`) fire as expected
- Able to control the component using `isShown`

**Review by joebingham-wk — DISMISSED** [review]
+1

File: lib/src/component/with_transition.dart [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+  /// key(s) in the `Map` are the same.
+  ///
+  /// __Example:__
+  ///
+  /// ```dart
+  /// UiFactory<WithTransitionExampleProps> WithTransitionExample = uiFunction(
+  ///   (props) {
+  ///     final isShown = useState(false);
+  ///
+  ///     (WithTransition()
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:52]: #nit this expression is unused;
```suggestion
  ///     return (WithTransition()
```

File: lib/src/component/with_transition.dart [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+  ///       ..childPropsByPhase = useMemo(() => {
+  ///         TransitionPhase.PRE_SHOWING: domProps()..className = 'before-showing',
+  ///         TransitionPhase.SHOWING: domProps()..className = 'showing',
+  ///         TransitionPhase.SHOWN: domProps()..className = 'shown',
+  ///         TransitionPhase.HIDING: domProps()..className = 'hiding',
+  ///         TransitionPhase.HIDDEN: domProps()..className = 'hidden',
+  ///       })
+  ///     )(
+  ///       // The child that has CSS transitions
+  ///     )
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:63]: #nit
```suggestion
  ///     );
```

File: lib/src/component/with_transition.dart [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+/// }
+///
+/// UiFactory<WithTransitionExampleProps> WithTransitionExample = uiFunction(
+///   (_props) {
+///     final defaultProps = useMemo(() {
+///       return WithTransitionExample()..initiallyShown = false;
+///     }, const []);
+///     final props = defaultProps..addProps(_props);
+///
+///     final isShown = useState(props.initiallyShown);
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:106]: #nit the preferred way to default props should be via null-awares (for a number of reasons, including promoting the values to non-nullable once nnbd is enabled)

```suggestion
///     final initiallyShown = props.initiallyShown ?? false;
///
///     final isShown = useState(initiallyShown);
```

File: lib/src/component/with_transition.dart [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+///       ..add('fade')
+///       ..add('in', isShown.value);
+///
+///     return Dom.div()(
+///       Dom.div()(
+///         (Dom.button()
+///           ..type = 'button'
+///           ..onClick = (useCallback((_) {
+///             isShown.setWithUpdater((isShown) => !isShown);
+///           }, const []))
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:118]: #nit unnecessary parentheses
```suggestion
///           ..onClick = useCallback((_) {
///             isShown.setWithUpdater((isShown) => !isShown);
///           }, const [])
```

File: lib/src/component/with_transition.dart [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+      ..childPropsByPhase = const {
+        v2.TransitionPhase.PRE_SHOWING: {},
+        v2.TransitionPhase.SHOWING: {},
+        v2.TransitionPhase.SHOWN: {},
+        v2.TransitionPhase.HIDING: {},
+        v2.TransitionPhase.HIDDEN: {},
+      }
+      ..addProps(v2.TransitionPropsMixin.defaultProps);
+    }, const []);
+    final props = defaultProps..addProps(_props);
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:184]: #nit prefer using null-awares to default props like mentioned above

File: lib/src/component/with_transition.dart [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+      transitionPhase.set(
+          _hasTransitionOut(tProps) ? TransitionPhase.HIDING : TransitionPhase.HIDDEN);
+    }
+
+    // getDerivedStateFromProps
+    if (props.isShown && _isOrWillBeHidden(transitionPhase.value)) {
+      _handleShow(props, transitionPhase.value);
+    } else if (!props.isShown && _isOrWillBeShown(transitionPhase.value)) {
+      _handleHide(props, transitionPhase.value);
+    }
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:215]: These calls mutating state and performing side-effects in `render`, which isn't allowed; they should go in `useEffect`

File: lib/src/component/with_transition.dart [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+        _isMounted.current = false;
+
+        _cancelTransitionEndTimer();
+        _cancelTransitionEventListener();
+      };
+    }, const []);
+
+    // componentDidUpdate
+    useLayoutEffect(() {
+      if (!_isMounted.current) return;
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:293]: Will this ever get called after the component is unmounted?

File: lib/src/component/with_transition.dart [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+          // triggers an update to `transitionPhase` that will result in this effect being processed.
+          break;
+        case TransitionPhase.HIDING:
+          // NOTE: props.onWillHide is called within _handleHide in the handler that
+          // triggers an update to `transitionPhase` that will result in this effect being processed.
+          _transitionNotGuaranteed.current = transitionPhase.value == TransitionPhase.SHOWING;
+          _handleHiding();
+          break;
+        case TransitionPhase.HIDDEN:
+          props.onDidHide?.call();
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:345]: I believe these need to also need to be included in the dependencies array: https://reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect

File: lib/src/component/with_transition.dart [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+    }
+
+    void _cancelTransitionEndTimer() {
+      _transitionEndTimer.current?.cancel();
+      _transitionEndTimer.current = null;
+    }
+
+    /// Listens for the next `transitionend` event and invokes a callback after
+    /// the event is dispatched.
+    void onNextTransitionEnd(Function() complete) {
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:249]: This references a number of props inside it (`props.transitionTimeout`, `_transitionInCount(props)`) that aren't declared in the calling `useLayoutEffect`

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
The hooks in this component are pretty complex; I'm see a couple issues, and then there are some areas in which I can't tell for sure whether rules of effect hooks are being broken.

It seems like this component's state logic doesn't translate super well to hooks, and that we'd need to rewrite it from the ground up with hooks in mind if we wanted to be confident in the implementation.

Perhaps for this time around, we should just use a class component under the hood, and expose a forwardRef component wrapper around it?

**Comment by aaronlademann-wf** [comment]
@greglittlefield-wf can you be more specific with what you're seeing as far as breaking the rules of hooks?

File: lib/src/component/with_transition.dart (hunk 12 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart:118]: The over_react analyzer plugin complains about this if you remove them.

**Comment by greglittlefield-wf** [comment]
>All tests we used for the old component are passing, and we have a consumer +10.

The current component has more edge cases and potentially incorrect code paths with the way it uses hooks, so I don't think that's enough to say the component in it's current state is bug-free.

Bugs introduced by improper uses of hooks might not always be surfaced in simple usages / tests, but rather by sequences of rerenders.

In this case, what about if you rerender the component when

File: lib/src/component/with_transition.dart (hunk 15 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart:293]: No, but it does get called once before the component mounts.

**Comment by aaronlademann-wf** [comment]
@greglittlefield-wf I have addressed your feedback.

File: lib/src/component/with_transition.dart (hunk 17 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart:234]: Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8

File: lib/src/component/with_transition.dart (hunk 16 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart:339]: Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8

File: lib/src/component/with_transition.dart (hunk 14 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart]: Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8

File: lib/src/component/with_transition.dart (hunk 13 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart]: Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8

File: lib/src/component/with_transition.dart (hunk 11 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart]: Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8

File: lib/src/component/with_transition.dart (hunk 10 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart]: Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8

File: lib/src/component/with_transition.dart (hunk 9 above)
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart]: Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8

File: lib/src/component/with_transition.dart [hunk 18]
```diff
@@ -0,0 +1,426 @@
...
+    _cancelTransitionEventListener();
+  }
+
+  @override
+  render() {
+    assert(_hasSingleValidChild(props));
+
+    final childElement = props.children.single as ReactElement;
+    final childProps = domProps({...getProps(childElement)});
+    final phaseProps = props.childPropsByPhase[state.$transitionPhase];
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:286]: While all of the methods this is being used in allow `null`, I think it'd be good to add a null-aware here to prevent null exceptions from occurring in the future
```suggestion
    final phaseProps = props.childPropsByPhase[state.$transitionPhase] ?? const {};
```

File: lib/src/component/with_transition.dart [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+    _cancelTransitionEndTimer();
+    _cancelTransitionEventListener();
+  }
+
+  @override
+  render() {
+    assert(_hasSingleValidChild(props));
+
+    final childElement = props.children.single as ReactElement;
+    final childProps = domProps({...getProps(childElement)});
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:285]: #nit unnecessary map copy
```suggestion
    final childProps = domProps(getProps(childElement));
```

File: lib/src/component/with_transition.dart [hunk 20]
```diff
@@ -0,0 +1,426 @@
...
+  get initialState => (newState()
+    ..$transitionPhase = _transitionPhaseDerivedFromProps(props)
+  );
+
+  @override
+  getDerivedStateFromProps(Map nextProps, Map prevState) {
+    final tNextProps = typedPropsFactory(nextProps);
+    final tPrevState = typedStateFactory(prevState);
+    if (tNextProps.isShown && _isOrWillBeHidden(tPrevState.$transitionPhase)) {
+      return _stateToBeginShowing(tNextProps, tPrevState.$transitionPhase);
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:221]: getDerivedSTateFromProps shouldn't have side effects (e.g., calling a props callback in this case) since in concurrent mode this method can be called multiple times without actually rendering the result. Can we move those to `componentDidUpdate`?

File: lib/src/component/with_transition.dart [hunk 21]
```diff
@@ -0,0 +1,426 @@
...
+      // changes waiting on `transitionend` events.
+      _cancelTransitionEventListener();
+    }
+
+    switch (state.$transitionPhase) {
+      case TransitionPhase.PRE_SHOWING:
+        _handlePreShowing();
+        break;
+      case TransitionPhase.SHOWING:
+        // NOTE: props.onWillShow is called within _handleShow in the handler that
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:246]: > _handleShow
This comment is out of date

File: lib/src/component/with_transition.dart [hunk 22]
```diff
@@ -0,0 +1,426 @@
...
+        // triggers an update to `$transitionPhase` that will result in this effect being processed.
+        break;
+      case TransitionPhase.HIDING:
+        // NOTE: props.onWillHide is called within _handleHide in the handler that
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:250]: > _handleHide
This comment is out of date

File: lib/src/component/with_transition.dart [hunk 23]
```diff
@@ -0,0 +1,426 @@
...
+
+    final childElement = props.children.single as ReactElement;
+    final childProps = domProps({...getProps(childElement)});
+    final phaseProps = props.childPropsByPhase[state.$transitionPhase];
+    final phaseClasses = ClassNameBuilder.fromProps(childProps)
+      ..addFromProps(phaseProps);
+
+    return cloneElement(childElement, domProps()
+      ..addTestId(childProps.getTestId())
+      ..addTestId('ovr.WithTransition.node')
```
Comment (greglittlefield-wf) [inline lib/src/component/with_transition.dart:292]: #supernit I like this abbreviation better: over_react -> o_r => or
```suggestion
      ..addTestId('or.WithTransition.node')
```

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]

File: lib/src/component/with_transition.dart [hunk 24]
```diff
@@ -0,0 +1,426 @@
...
+        _handlePreShowing();
+        break;
+      case TransitionPhase.SHOWING:
+        // NOTE: props.onWillShow is called within _handleShow in the handler that
+        // triggers an update to `$transitionPhase` that will result in this effect being processed.
+        break;
+      case TransitionPhase.HIDING:
+        // NOTE: props.onWillHide is called within _handleHide in the handler that
+        // triggers an update to `$transitionPhase` that will result in this effect being processed.
+        _transitionNotGuaranteed = state.$transitionPhase == TransitionPhase.SHOWING;
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart:252]: If I change it to `tPrevState` - two tests fail.

[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/component/with_transition.dart:250]: So perhaps this flag isn't needed anymore? If the code isn't changed to prevState, then the flag should be eliminated, because right now it's always `false`.

Do the AbstractTransition tests fail when that code I linked is changed to `_transitionNotGuaranteed = false;`?

[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart:250]: Yes

[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/component/with_transition.dart:250]: Cool, sounds like this impl just doesn't need it, likely because of the different way state transitions work

[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/with_transition.dart]: Removed in https://github.com/Workiva/over_react/pull/621/commits/e0208a6375576540b0fd0cccc5ffd2e5cc74f611

**Review by greglittlefield-wf — APPROVED** [review]

**Review by joebingham-wk — APPROVED** [review]

**Comment by maxwellpeterson-wf** [comment]
+10
- pulled into https://github.com/Workiva/wdesk_sdk/pull/5096 where there are two separate usages of `WithTransition`. Both work as expected.

**Comment by greglittlefield-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
